### ISSUE TITLE
T440s with kernel >= 3.15 doesn't power off properly  (Analysis and possible solution included !)

### ISSUE THREAD

**Issue by smunaut** [Bumblebee-Project/bbswitch#112]
Hi,

So I'm using a T440s with a modern kernel that reports "Windows 2013" compatibility (and soon Windows 2015 with kernel 4.2). This breaks bbswitch because of some changes this triggers in the ACPI table. Manually overriding acpi_osi does fix the issue and allow the current bbswitch to work, but what I'm looking it here is how to make it work with the "new" Win 8.1 method of shutting down the card.

So the main symptoms of the issue are :

 * Using vanilla bbswith : Switching OFF does put it to D3, but it's not entirely off. The power usage is lower but not as low as it should be (in my case, about 1W lower than full power while it should be more like 2.5W). Also, bbswith reports the card as "ON" because lspci still works fine and you can still access the config space. There is no way to turn the card back ON and so loading nvidia driver will actually crash trying to access a card that's only half powered.

 * Using the bbswitch patched like in PR #102 . This is the same effect, power down is not complete and you only save 1W instead of 2.5W. But at least you can turn the card ON and OFF properly. But once nvidia has been loaded, then it won't switch off at all ....

This is the DSDT table from the T440s with the latest bios (which even has Windows 10 support) : 
http://pastebin.com/raw.php?i=C6Q3A8aa

The important thing to note is that when "Windown 2013" string is found, then OSYS is set to 0x07DD. This in turn cause VMSH to be set to 1. This in turn causes \_SB_.PCI0.PEG_.VID_._PS3 to _NOT_ call GPOF ... and so the card is never really turned off completely.

Now if you look at how GPOF can be called, you can see it will be called as part of NVP3 power resource which is _PR3 ... but on the node \_SB_.PCI0.PEG_   and not \_SB_.PCI0.PEG_.VID_ !

So basically you need to put the PCIe root port (parent pci device) in D3 and not just the card.

I tested this and it indeed triggered the proper expected power saving and seemed to behave exactly like if I tweaked acpi_osi.

**Comment by doudou** [comment]
Do you have code to put the PCIe root in D3 as well ? Would gladly integrate it in #102.

**Comment by smunaut** [comment]
This is what I've been using this last week on my T440s : 

https://github.com/smunaut/bbswitch/commit/ca87980cb6105c34e1a138b553eac602d8151519

Note that this is non-conditional ... maybe there should be a whitelist of supported machines ?
I know on the T440s, the subsystem product id changes depending on which mode (i.e. old mode or new mode) is currently active. (So if you boot with an old kernel you see one subsystem id and if you boot with a new one, you see another).

**Comment by doudou** [comment]
@smunaut: this change causes memory corruption after I wake up the NVidia card (i.e. back to base one). I am really not a kernel developer, but the obvious problem is that the patch puts the port in D3 but the actual PCIe driver is not informed of that.

I've checked the kernel source code, PCIe ports do not seem to have runtime suspend/resume. That would probably be the better fix (use a dummy nvidia driver to suspend/resume the NV card using the kernel PM handlers and have the PCIe root port go into suspend accordingly). I already wrote the dummy driver, but I'm now trying to find someone I could contact that would be willing to look into this. If you know anybody doing kernel dev ...

The alternative (that I'm going to try as soon as I have the time) would be to unbind the PCIe root port driver from the port and bind yet-another-dummy-driver to put the port to sleep.

**Comment by smunaut** [comment]
@doudou Mmm ... when I think about it, it must be #78 ... I mean for all intents and purposes, my patch will trigger the same behavior in ACPI as when using the acpi_osi string to revert to the old shutdown behavior.

AFAIK there is no drivers binded to the pcie ports themselves, which is why they don't do auto runtime/suspend. I did raise the issue on linux-pm and linux-acpi but didn't really get any answers.

**Comment by doudou** [comment]
> AFAIK there is no drivers binded to the pcie ports themselves

It seems that there is. lspci reports "pcieport", which matches the driver in drivers/pci/pcie/portdrv_pci.c.

As to whether we can unbind it, that's another story altogether ;-)

**Comment by doudou** [comment]
> As to whether we can unbind it, that's another story altogether ;-)

Yes, it seems to work

After

~~~
echo 0000:00:01.1 > bind 
~~~

the driver is not listed as driver of the port, and after

~~~
echo 0000:00:01.1 > unbind 
~~~

it gets listed again

**Comment by smunaut** [comment]
I'd also point out that doing the ACPI DSM call is useless with this new method. It's not required and so maybe removing it would help your issues ?
Did you try shutdown when pcieport is unbinded ? Did it help ?

**Comment by doudou** [comment]
Yes, I already removed the DSM calls.

I'm going to push the module I wrote, which basically registers a PCI driver for the NV card and for the PCIe root port, and manages to make them autosuspend. They both go to D3 on autosuspend, and then boom, the system crashes hard. I did not investigate further, I should really be doing something else than this ;-)

I start to believe that the corruption issues directly stem from the hard shutdown of the NV card.

**Comment by doudou** [comment]
Some progress.

I've managed to get to a state where both the NVidia and PCIe port get into suspend, using *only* normal runtime suspend kernel paths. Basically, they get into D3 thanks to the common PCI power-management code. The code is on [this branch](https://github.com/doudou/bbswitch/tree/runtime_suspend).

The catch: I get the memory corruption anyways as soon as the card(s) are woken up :( I want to do some reading in intel documentation, whether there are things that need to be done before one is allowed to put a PCIe port in D3.

Note that because I use runtime suspend, you *cannot* use lspci to check the card's state. This wakes it up. I've used systemtap and powertop to verify that (1) the kernel was attempting to put the cards in D3 and succeeded, and (2) that it led to significant power savings (it did, almost 2W more than with having the NV card in D3).

**Comment by gsgatlin** [comment]
@doudou Does this code mess up non thinkpad laptops? Like do you plan on doing a pull request?

**Comment by doudou** [comment]
> Does this code mess up non thinkpad laptops?

Don't know ... only have a thinkpad

> Like do you plan on doing a pull request?

No, given its effects on mine ... I already have a pull request (#102) on a method that kind-of works. Meaning:
 - has some power benefits (put the NV card in D3, but not the root port) without filesystem corruption
 - is compatible with the bumblebee daemon
 - but is unable to put the card back to sleep after the nvidia module got loaded

I personally use this one. Left the rest on the side, I don't have time for this right now.

**Comment by gsgatlin** [comment]
@doudou Ok. Cool. Thanks for the info. If you ever have time to get it in a state where you'd like it to be tested on a non thinkpad I have a spare optimus ideapad I could test it on.

**Comment by sharms** [comment]
I successfully tested the patch on Fedora 23 / 4.2.1-300 on a Thinkpad W541 / Optimus K2100 and power consumption went from 25W to 19W.  Thanks!

**Comment by storrgie** [comment]
Is there a plan to merge this fix into mainline? I'm about to jump to F23.

**Comment by Lacrymology** [comment]
I'm seeing some possibly related issue with a brand new T450s

**Comment by storrgie** [comment]
It'd be very nice if we could get this mainlined rather than have it as a required patch.

**Comment by Lacrymology** [comment]
Sorry, can someone point me to how to recompile this and try it?

**Comment by smunaut** [comment]
@storrgie The issue is that this fix is completely ThinkPad specific ... It would probably break things on other brands.

(and even on thinkpads it's only been tested on a few models)

**Comment by Lacrymology** [comment]
ok, that was a silly question, 

but I'm getting the following:

```
lacrymology@hex ~/aur/bbswitch $ make
make -C /lib/modules/4.2.1-1-ARCH/build M="$(pwd)" modules
make[1]: Entering directory '/usr/lib/modules/4.2.1-1-ARCH/build'
make[1]: *** No rule to make target 'modules'.  Stop.
make[1]: Leaving directory '/usr/lib/modules/4.2.1-1-ARCH/build'
Makefile:12: recipe for target 'default' failed
make: *** [default] Error 2
```

**Comment by Lacrymology** [comment]
this is still failing for me, I'm getting this when trying to load bbswitch

```
[  454.627938] bbswitch: Found integrated VGA device 0000:00:02.0: \_SB_.PCI0.VID_
[  454.627945] bbswitch: Found discrete VGA device 0000:04:00.0: \_SB_.PCI0.PEG_.VID_
[  454.627957] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
[  454.629069] bbswitch: detected an Optimus _DSM function
[  454.629872] bbswitch: Succesfully loaded. Discrete card 0000:04:00.0 is on
```

**Comment by Lacrymology** [comment]
trying to turn off with tee:

```
# tee /proc/acpi/bbswitch<<<OFF
# dmesg
...
[  531.504356] bbswitch: device 0000:04:00.0 is in use by driver 'bbswitch_nv', refusing OFF
# rmmod bbswitch_nv
# dmesg
[  546.959043] bbswitch_nv: attempting to resume
[  546.959058] bbswitch_nv: attempting to suspend
[  547.076317] bbswitch_nv: attempting to resume
# tee /proc/acpi/bbswitch<<<OFF
# dmesg
[  553.644208] bbswitch: disabling discrete graphics
[  553.644225] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
[  553.659670] pci 0000:04:00.0: Refused to change power state, currently in D0
```

**Comment by Lacrymology** [comment]
bumblebee works correctly, in the sense of loading the nvidia driver when run with optirun, it's just that the card won't switch off

**Comment by Lacrymology** [comment]
here's my lspci -v output for the nvidia card (while running glxgears with optirun)

```
04:00.0 3D controller: NVIDIA Corporation GM108M [GeForce 940M] (rev a2)
        Subsystem: Lenovo Device 5037
        Flags: bus master, fast devsel, latency 0, IRQ 49
        Memory at f1000000 (32-bit, non-prefetchable) [size=16M]
        Memory at c0000000 (64-bit, prefetchable) [size=256M]
        Memory at d0000000 (64-bit, prefetchable) [size=32M]
        I/O ports at 3000 [size=128]
        Capabilities: [60] Power Management version 3
        Capabilities: [68] MSI: Enable+ Count=1/1 Maskable- 64bit+
        Capabilities: [78] Express Endpoint, MSI 00
        Capabilities: [100] Virtual Channel
        Capabilities: [250] Latency Tolerance Reporting
        Capabilities: [258] L1 PM Substates
        Capabilities: [128] Power Budgeting <?>
        Capabilities: [600] Vendor Specific Information: ID=0001 Rev=1 Len=024 <?>
        Kernel driver in use: nvidia
        Kernel modules: nouveau, nvidia
```

**Comment by Lacrymology** [comment]
dmesg when running optirun:

```
[ 1022.584913] NVLINK: Nvlink Core is being initialized, major device number 244
[ 1022.585100] [drm] Initialized nvidia-drm 0.0.0 20150116 for 0000:04:00.0 on minor 1
[ 1022.585108] NVRM: loading NVIDIA UNIX x86_64 Kernel Module  355.11  Wed Aug 26 16:35:41 PDT 2015
[ 1022.829610] vgaarb: this pci device is not a vga device
[ 1022.833781] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
[ 1022.834091] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
[ 1022.834244] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
[ 1022.834389] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
[ 1022.834948] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
[ 1022.835382] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
[ 1022.835532] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
[ 1022.873564] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
[ 1023.067067] vgaarb: this pci device is not a vga device
```

and when stopping it:

```
[ 1065.933239] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
[ 1065.949611] NVLINK: Unregistered the Nvlink Core, major device number 244
[ 1065.951791] [drm] Module unloaded
[ 1065.968420] bbswitch: disabling discrete graphics
[ 1065.968435] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
[ 1065.980288] pci 0000:04:00.0: Refused to change power state, currently in D0
```

**Comment by smunaut** [comment]
There is _nothing_ that indicates in your logs that it's not working.

The "Refused to change power state" is inconsequential.

The only way to check if it's working or not is to check if the rev is "ff" during a lspci when it's turned off.

**Comment by Lacrymology** [comment]
@smunaut the rev is not ff, it doesn't change. The first time I tried to turn it off with the vanilla module I got a traceback, as well, I'd forgotten about that:

```
[  163.280190]  snd_pcm i2c_algo_bit i2c_i801 ptp shpchp snd_timer snd soundcore pps_core lpc_ich processor sch_fq_codel ip_tables x_tables ext4 crc16 mbcache jbd2 sd_mod rtsx_pci_sdmmc mmc_core atkbd libps2 ahci libahci libata scsi_mod xhci_pci ehci_pci ehci_hcd xhci_hcd rtsx_pci usbcore usb_common i8042 serio
[  163.280205] CPU: 0 PID: 2385 Comm: tee Tainted: G        W  O    4.2.1-1-ARCH #1
[  163.280206] Hardware name: LENOVO 20BXCTO1WW/20BXCTO1WW, BIOS JBET51WW (1.16 ) 07/08/2015
[  163.280207]  0000000000000000 00000000619135a4 ffff88030e58fcc8 ffffffff8156b77a
[  163.280209]  0000000000000000 ffff88030e58fd20 ffff88030e58fd08 ffffffff81074846
[  163.280210]  ffff88030e58fce8 ffff88033f6e1000 ffff88033f661c40 00007ffdb03cc1a0
[  163.280212] Call Trace:
[  163.280216]  [<ffffffff8156b77a>] dump_stack+0x4c/0x6e
[  163.280218]  [<ffffffff81074846>] warn_slowpath_common+0x86/0xc0
[  163.280220]  [<ffffffff810748d5>] warn_slowpath_fmt+0x55/0x70
[  163.280223]  [<ffffffff812f503b>] ? __pci_set_master+0x3b/0xf0
[  163.280224]  [<ffffffff812f7b50>] pci_disable_device+0xb0/0xd0
[  163.280227]  [<ffffffffa07714ad>] bbswitch_off+0xad/0x240 [bbswitch]
[  163.280228]  [<ffffffffa077187b>] bbswitch_proc_write+0x9b/0xb2 [bbswitch]
[  163.280231]  [<ffffffff81236742>] proc_reg_write+0x42/0x70
[  163.280233]  [<ffffffff811d01c7>] __vfs_write+0x37/0x100
[  163.280235]  [<ffffffff811d2fd8>] ? __sb_start_write+0x58/0x100
[  163.280237]  [<ffffffff811d0ad4>] vfs_write+0xa4/0x1a0
[  163.280238]  [<ffffffff811d0a14>] ? vfs_read+0x114/0x130
[  163.280240]  [<ffffffff811d17e5>] SyS_write+0x55/0xc0
[  163.280242]  [<ffffffff81570cee>] entry_SYSCALL_64_fastpath+0x12/0x71
[  163.280243] ---[ end trace be475d6cb25e4d1e ]---
```

Also, according to KDE's battery status panel, I do get an hour less of battery while optirun is running something, but when it's supposed to be turned off, I'm still getting this with lspci

```
04:00.0 3D controller: NVIDIA Corporation GM108M [GeForce 940M] (rev a2)
```

Notice that as reported above by other people, I'm also getting 

```
root@hex /home/lacrymology/aur/bbswitch # cat /proc/acpi/bbswitch 
0000:04:00.0 ON
```

regardless

**Comment by smunaut** [comment]
Are you sure you took the source from my repo and switched to the right branch ?  (the default at checkout is 'master' and doesn't have the patch)

**Comment by Lacrymology** [comment]
yes, branch `runtime_suspend`, right? I even copied the module over to `/lib/modules/...` manually and tried to load it with modprobe, it lodaed correctly (after running `depmod -a`), but still the same results:

```
# dmesg
[ 1256.134906] bbswitch: version 0.8
[ 1256.134918] bbswitch: Found integrated VGA device 0000:00:02.0: \_SB_.PCI0.VID_
[ 1256.134926] bbswitch: Found discrete VGA device 0000:04:00.0: \_SB_.PCI0.PEG_.VID_
[ 1256.134938] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
[ 1256.135548] bbswitch: detected an Optimus _DSM function
[ 1256.135564] bbswitch: Succesfully loaded. Discrete card 0000:04:00.0 is on
[ 1282.007356] bbswitch: disabling discrete graphics
[ 1282.007365] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
# lspci
04:00.0 3D controller: NVIDIA Corporation GM108M [GeForce 940M] (rev a2)
        Subsystem: Lenovo Device 5037
        Flags: fast devsel, IRQ 16
        Memory at f1000000 (32-bit, non-prefetchable) [size=16M]
        Memory at c0000000 (64-bit, prefetchable) [size=256M]
        Memory at d0000000 (64-bit, prefetchable) [size=32M]
        I/O ports at 3000 [size=128]
        Expansion ROM at <ignored> [disabled]
        Capabilities: [60] Power Management version 3
        Capabilities: [68] MSI: Enable- Count=1/1 Maskable- 64bit+
        Capabilities: [78] Express Endpoint, MSI 00
        Capabilities: [100] Virtual Channel
        Capabilities: [250] Latency Tolerance Reporting
        Capabilities: [258] L1 PM Substates
        Capabilities: [128] Power Budgeting <?>
        Capabilities: [600] Vendor Specific Information: ID=0001 Rev=1 Len=024 <?>
        Capabilities: [900] #19
        Kernel modules: nouveau, nvidia
```

**Comment by Lacrymology** [comment]
no, wait, I'm using @doudou 's code. I'll try yours now

**Comment by Lacrymology** [comment]
@smunaut your code reports `rev: ff` in lspci, and running with glxgears turns`optirun` turns it back on successfully! 

Is there any power consumption metering app that can help me confirm exactly how much power is getting sucked?

**Comment by smunaut** [comment]
powertop  but it only reports it when on battery (also you need to wait a few minutes for it to stabilise)

**Comment by Lacrymology** [comment]
12.4 W with the card on, 8.6 with the card off. No idea if it's about the right amount (nvidia 940M) but it's definitely something

**Comment by doudou** [comment]
@smunaut: for what it's worth, I had trouble with your patch a while ago (crashed). I updated my bios a few days ago and just tried it again. It works ! Perfectly ! Thank you so much !

**Comment by smunaut** [comment]
@doudou Nice :)
@Lacrymology  Yeah ~3-4W sounds about right.

**Comment by Lacrymology** [comment]
I know next to nothing about hardware handling, but isn't there a flag or
id that can be read to make this if-able? I guess it'd come down to a
whitelist if it's to be automatable, but I'm thinking it could be made a
module parameter

**Comment by smunaut** [comment]
You could probab know if this new method should be used by looking at the PCI subdevice IDs. The ACPI table dump of my laptop shows that the subdevices IDs are changed depending on wether the new method is applied or not.

**Comment by Lacrymology** [comment]
but, again, are they changed in a generic way (i.e. a testable mask), or would we need to build a whitelist of ids which require the patch? How do I get an ACPI table dump of my laptop to compare? Or.. well, are you interested in persuing this? I'm willing to help, I'm a programmer, but I've no experience in hardware/drivers/kernels

**Comment by chrk123** [comment]
@smunaut I've compiled your bbswitch hack-t440s module on my t440p. So far, according to lspci, it seems to work. 

When /proc/acpi/bbswitch states that the card is **off**, lspci reports:
`02:00.0 VGA compatible controller: NVIDIA Corporation GK208M [GeForce GT 730M] (rev ff)`

and when /proc/acpi/bbswitch states that the card is **on**, lspci reports: 
`02:00.0 VGA compatible controller: NVIDIA Corporation GK208M [GeForce GT 730M] (rev a1)`

However, the battery discharge rate stays the same at about 12W to 13W. Furthermore, the temperature readings from my notebook components didn't changed either (55°C). I didn't gave power consumption much thought during the last few months, but normally, I had about <=8W discharge rate in idle and a system temperature below 50°C, which let's me assume that my discrete graphics card is still not properly shut down. 

I'm using a T440p , Linux t440p 4.2.3-1-ARCH #1 SMP PREEMPT Sat Oct 3 18:52:50 CEST 2015 x86_64 GNU/Linux on Arch-Linux

lspci -v when turned on:

    02:00.0 VGA compatible controller: NVIDIA Corporation GK208M [GeForce GT 730M] (rev a1)         prog-if 00 [VGA controller])
	Subsystem: Lenovo Device 221d
	Flags: bus master, fast devsel, latency 0, IRQ 17
	Memory at f0000000 (32-bit, non-prefetchable) [size=16M]
	Memory at c0000000 (64-bit, prefetchable) [size=256M]
	Memory at d0000000 (64-bit, prefetchable) [size=32M]
	I/O ports at 3000 [size=128]
	Capabilities: [60] Power Management version 3
	Capabilities: [68] MSI: Enable- Count=1/1 Maskable- 64bit+
	Capabilities: [78] Express Endpoint, MSI 00
	Capabilities: [100] Virtual Channel
	Capabilities: [128] Power Budgeting <?>
	Capabilities: [600] Vendor Specific Information: ID=0001 Rev=1 Len=024 <?>
	Capabilities: [900] #19
	Kernel modules: nouveau, nvidia

lspci -v when turned off

    02:00.0 VGA compatible controller: NVIDIA Corporation GK208M [GeForce GT 730M] (rev ff) (prog-if ff)
    !!! Unknown header type 7f
    Kernel modules: nouveau, nvidia

dmesg seems fine when using echo OFF/ON > /proc/acpi/bbswitch

    [ 2583.030756] bbswitch: disabling discrete graphics
    [ 2583.030766] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found 
    [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
    [ 2961.430922] bbswitch: enabling discrete graphics


The only thing that is suspicious is the dmesg output after running optirun:


    [ 3176.964358] bbswitch: disabling discrete graphics
    [ 3176.964369] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch -Found [Buffer], ACPI requires [Package] (20150619/nsarguments-95)
    [ 3176.976934] pci 0000:02:00.0: Refused to change power state, currently in D0


which is currently my only hint but you said it's inconsequantial. Does anyone have further hints for me? Thanks!

**Comment by emanuil-tolev** [comment]
@smunaut have you tried / had any luck with turning off the card completely on your Thinkpad? Unfortunately it seems that the Windows 2013 workaround causes FS corruption after resuming from sleep (even described by a commenter above). I presumed that uninstalling+purging bumblebee and everything nvidia related and rmmod-ing nouveau would be sufficient to turn it off as the kernel just wouldn't turn it on.

**Comment by doudou** [comment]
Hi @emanuil-tolev: I'm running @smunaut version of the bumblebee module on a Thinkpad t440p successfully. Needed to get the latest BIOS though. Tests done with previous BIOS versions were causing crashes/corruption.

**Comment by emanuil-tolev** [comment]
Alright, thanks @doudou - I'll try that tomorrow (just about to unhappily run out of a 96Wh battery on a T450s due to this :)).

**Comment by emanuil-tolev** [comment]
Following the instructions in https://github.com/smunaut/bbswitch/tree/hack-t440s did work for my Thinkpad T450s.

After loading the module (just simple load, I'll install it using dkms later):

```
emanuil@midori:~/software/bbswitch$ sudo make load
rmmod bbswitch
rmmod: ERROR: Module bbswitch is not currently loaded
make: [load] Error 1 (ignored)
insmod bbswitch.ko

emanuil@midori:~/software/bbswitch$ cat /proc/acpi/bbswitch
0000:04:00.0 ON

emanuil@midori:~/software/bbswitch$ dmesg | tail
[ 5618.042276] bbswitch: module verification failed: signature and/or  required key missing - tainting kernel
[ 5618.042676] bbswitch: version 0.8
[ 5618.042684] bbswitch: Found integrated VGA device 0000:00:02.0: \_SB_.PCI0.VID_
[ 5618.042695] bbswitch: Found discrete VGA device 0000:04:00.0 (on 0000:00:1c.4): \_SB_.PCI0.PEG_.VID_
[ 5618.042710] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20140424/nsarguments-95)
[ 5618.043515] bbswitch: detected an Optimus _DSM function
[ 5618.043590] bbswitch: Succesfully loaded. Discrete card 0000:04:00.0 is on
```

After instructing it to turn the card off:

```
emanuil@midori:~/software/bbswitch$ sudo tee /proc/acpi/bbswitch <<<OFF
OFF

emanuil@midori:~/software/bbswitch$ lspci -v | grep -i -A 10 nvidia
04:00.0 3D controller: NVIDIA Corporation GM108M [GeForce 940M] (rev ff) (prog-if ff)
        !!! Unknown header type 7f

emanuil@midori:~/software/bbswitch$ cat /proc/acpi/bbswitch
0000:04:00.0 OFF

emanuil@midori:~/software/bbswitch$ dmesg | tail
[ 5658.848202] bbswitch: disabling discrete graphics
[ 5658.848225] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20140424/nsarguments-95)
[ 5658.878253] thinkpad_acpi: EC reports that Thermal Table has changed
```

**Power consumption falls by ~2W (powertop).**

After turning the card back ON with bbswitch, I get this output:

```
emanuil@midori:~/software/bbswitch$ sudo tee /proc/acpi/bbswitch <<<ON
ON

emanuil@midori:~/software/bbswitch$ dmesg | tail -2
[10178.009129] bbswitch: enabling discrete graphics
[10178.113447] thinkpad_acpi: EC reports that Thermal Table has changed

emanuil@midori:~/software/bbswitch$ cat /proc/acpi/bbswitch
0000:04:00.0 ON

emanuil@midori:~/software/bbswitch$ lspci -v | grep -i -A 10 nvidia
04:00.0 3D controller: NVIDIA Corporation GM108M [GeForce 940M] (rev a2)
        Subsystem: Lenovo Device 5037
        Flags: bus master, fast devsel, latency 0, IRQ 16
        Memory at f3000000 (32-bit, non-prefetchable) [size=16M]
        Memory at e0000000 (64-bit, prefetchable) [size=256M]
        Memory at f0000000 (64-bit, prefetchable) [size=32M]
        I/O ports at 3000 [size=128]
        Expansion ROM at <ignored> [disabled]
        Capabilities: <access denied>
```

and power consumption climbs up by about 2W again. I actually expected a bit more than that, but it seems that the Chrome browser and possibly the wifi drivers are both performing significantly worse (in terms of power) than on Windows. But that's out of bbswitch's scope. Sometimes varying system load makes it seem like the power consumption has barely changed by 1W or less - I had to actually stop doing anything, sit down and watch powertop for a few minutes to identify the effect (it is there). It does also make a long-term difference to the battery life, on average it does last 1-2h longer now.

No data corruption (I've slept and restored a few times with the card off) so far.

**Comment by smunaut** [comment]
Yeah powertop needs a few minutes. that's because it "deduces" the power consumptions by looking at how fast the battery discharges AFAIK and it averages the results over a few minutes.

2W is about what I get too on a T440s.  10-11W -> 8-9W.

**Comment by hgomersall** [comment]
I'm experiencing a similar problem on a t450s - hack-t440s seems to make bbswitch work properly, though it requires a manual unloading of the drivers first. At risk of polluting the thread, where should the unloading occur - so I can go and ask the right question? (bumblebee seems to load the modules just fine, but never unloads them.)

**Comment by michaeldfallen** [comment]
I'm on a T450s trying the hack-t440s branch. Running `sudo tee /proc/acpi/bbswitch <<<OFF` results in this message in dmesg `pci 0000:04:00.0: disabling already-disabled device`:
```
[  255.022982] ------------[ cut here ]------------
[  255.022988] WARNING: CPU: 2 PID: 3239 at /build/linux-jJX2Dm/linux-4.2.0/drivers/pci/pci.c:1550 pci_disable_device+0xb0/0xd0()
[  255.022990] pci 0000:04:00.0: disabling already-disabled device
[  255.022991] Modules linked in:
[  255.022992]  rfcomm ctr ccm msr acpi_call(OE) xt_addrtype xt_conntrack ipt_MASQUERADE nf_nat_masquerade_ipv4 iptable_nat nf_conntrack_ipv4 nf_defrag_ipv4 nf_nat_ipv4 iptable_filter ip_tables x_tables nf_nat nf_conntrack bridge stp llc aufs pci_stub vboxpci(OE) vboxnetadp(OE) vboxnetflt(OE) vboxdrv(OE) bbswitch(OE) bnep uvcvideo videobuf2_vmalloc videobuf2_memops videobuf2_core v4l2_common videodev media nls_iso8859_1 btusb btrtl btbcm btintel bluetooth arc4 intel_rapl iosf_mbi x86_pkg_temp_thermal intel_powerclamp coretemp snd_hda_codec_hdmi snd_hda_codec_realtek snd_hda_codec_generic kvm_intel kvm snd_hda_intel snd_hda_codec snd_hda_core iwlmvm snd_hwdep joydev mac80211 input_leds serio_raw snd_seq_midi snd_seq_midi_event snd_pcm snd_rawmidi thinkpad_acpi iwlwifi nvram snd_seq rtsx_pci_ms snd_seq_device
[  255.023041]  cfg80211 snd_timer lpc_ich memstick mei_me snd mei shpchp soundcore wmi mac_hid nfsd auth_rpcgss nfs_acl lockd grace sunrpc parport_pc ppdev lp parport autofs4 jitterentropy_rng drbg ansi_cprng algif_skcipher af_alg dm_crypt rtsx_pci_sdmmc crct10dif_pclmul crc32_pclmul ghash_clmulni_intel aesni_intel i915 aes_x86_64 lrw gf128mul glue_helper ablk_helper cryptd i2c_algo_bit psmouse drm_kms_helper ahci e1000e libahci drm ptp rtsx_pci pps_core video
[  255.023062] CPU: 2 PID: 3239 Comm: tee Tainted: G        W  OE   4.2.0-18-generic #22-Ubuntu
[  255.023063] Hardware name: LENOVO 20BXCTO1WW/20BXCTO1WW, BIOS JBET53WW (1.18 ) 09/14/2015
[  255.023065]  0000000000000000 0000000054022c85 ffff8800a86ebd48 ffffffff817e8c09
[  255.023067]  0000000000000000 ffff8800a86ebda0 ffff8800a86ebd88 ffffffff8107b3c6
[  255.023069]  ffff8800a86ebd68 ffff8805354c9000 ffff88053544dc40 ffff8800a86ebf18
[  255.023071] Call Trace:
[  255.023076]  [<ffffffff817e8c09>] dump_stack+0x45/0x57
[  255.023079]  [<ffffffff8107b3c6>] warn_slowpath_common+0x86/0xc0
[  255.023081]  [<ffffffff8107b455>] warn_slowpath_fmt+0x55/0x70
[  255.023084]  [<ffffffff8140fd4b>] ? __pci_set_master+0x3b/0xf0
[  255.023086]  [<ffffffff81412870>] pci_disable_device+0xb0/0xd0
[  255.023090]  [<ffffffffc07894ad>] bbswitch_off+0xad/0x240 [bbswitch]
[  255.023092]  [<ffffffffc078987b>] bbswitch_proc_write+0x9b/0xb2 [bbswitch]
[  255.023103]  [<ffffffff81266872>] proc_reg_write+0x42/0x70
[  255.023106]  [<ffffffff811fc878>] __vfs_write+0x18/0x40
[  255.023107]  [<ffffffff811fd206>] vfs_write+0xa6/0x1a0
[  255.023109]  [<ffffffff811fdef5>] SyS_write+0x55/0xc0
[  255.023112]  [<ffffffff817ef9f2>] entry_SYSCALL_64_fastpath+0x16/0x75
[  255.023114] ---[ end trace 215be7042c35cc92 ]---
```

Result of `lscpi`:
```
04:00.0 3D controller: NVIDIA Corporation GM108M [GeForce 940M] (rev a2)
```

So the device is powered or partially powered off but not in full `rev ff`?

**Comment by Lacrymology** [comment]
since the latest update to bbswitch/bumblebee once the nvidia module is loaded to the kernel I can't turn it off anymore, and the device can't be turned off either..

**Comment by Lacrymology** [comment]
I think I might be seeing the same issue @hgomersall is referring to, although I haven't been able to manually remove the module either

**Comment by Lacrymology** [comment]
ok, I had to:

```bash
modprobe -r nvidia_modeset
modprobe -r nvidia
tee /proc/acpi/bbswitch <<< OFF
```

for it to work

**Comment by ArchangeGabriel** [comment]
@hgomersall @Lacrymology That’s https://github.com/Bumblebee-Project/Bumblebee/issues/699.

@smunaut @doudou There is clearly space for improvement, especially regarding declared ACPI_OSI and their impact on ASL functions. We need to make some summary of everything we know about this currently, what we should test to go further and then probably rewrite part of bbswitch/kernel code according to this. Maybe linux-pm or -acpi MLs are better places for this, and people like Rafael Wysocki might be interested in this.

**Comment by smunaut** [comment]
One thing I've noticed is that the subdevice VID:PID change when the new method is enabled in the ACPI DSDT so maybe it could be based on that to select one method or the other.

**Comment by ArchangeGabriel** [comment]
Yes, but it might be different for other laptops. Are all the (currently-known) affected laptop following this rule?

Is there any non-Thinkpad laptop concerned? What about Thinkpad laptop? I suppose this need a recent BIOS expecting `Windows 2013` OSI?

**Comment by ArchangeGabriel** [comment]
Also, 3.15 is indeed the culprit because of https://github.com/torvalds/linux/commit/faae404ebdc6bba744919d82e64c16448eb24a36.

**Comment by smunaut** [comment]
Then a bit later : https://github.com/torvalds/linux/commit/796888e942b34cbbd738d9e5478b7d103ee38061#diff-aa93b5317c200560767b97a9d9301bd8

**Comment by ArchangeGabriel** [comment]
Yes, so this probably explain that we have a lot of people reporting issue after going from 3.14 to 3.16 and other ones after going to 4.1.

**Comment by klebed** [comment]
Does that mean, that lenovo bios disabling something, if we are not telling him, that we are booting windows 8 or something else happened?

**Comment by ArchangeGabriel** [comment]
Yes, the BIOS disable some functions if you don’t tell it that you’re booting the latest Windows.

**Comment by klebed** [comment]
And then the solution is to add acpi_osi="Windows 2013" in kernel boot parameters? I've seen some reports, that people getting fs corruption and crashes then.
Is it worth to do?

**Comment by ArchangeGabriel** [comment]
This is automatically added since kernel 3.15. So you should either add acpi_osi="!Windows 2013" (not recommended) or try @smunaut patch (look at https://github.com/Bumblebee-Project/bbswitch/issues/112#issuecomment-124669413).

**Comment by klebed** [comment]
Ok... I've tested hacked version of bbswitch. At least now it seems like working.

Here is the logs, while running glxgears with optirun and exiting then:
[   17.180425] thinkpad_ec: thinkpad_ec_request_row: arg0 rejected: (0x01:0x00)->0x00
[   17.180429] thinkpad_ec: thinkpad_ec_read_row: failed requesting row: (0x01:0x00)->0xfffffffb
[   17.180431] thinkpad_ec: initial ec test failed
[   17.215779] thinkpad_ec: thinkpad_ec_request_row: arg0 rejected: (0x01:0x00)->0x00
[   17.215782] thinkpad_ec: thinkpad_ec_read_row: failed requesting row: (0x01:0x00)->0xfffffffb
[   17.215784] thinkpad_ec: initial ec test failed
[  116.236037] bbswitch: enabling discrete graphics
[  116.508646] thinkpad_acpi: EC reports that Thermal Table has changed
[  116.585041] nvidia: module license 'NVIDIA' taints kernel.
[  116.585045] Disabling lock debugging due to kernel taint
[  116.594804] vgaarb: device changed decodes: PCI:0000:04:00.0,olddecodes=io+mem,decodes=none:owns=none
[  116.595109] [drm] Initialized nvidia-drm 0.0.0 20150116 for 0000:04:00.0 on minor 1
[  116.595114] NVRM: loading NVIDIA UNIX x86_64 Kernel Module  340.96  Sun Nov  8 22:33:28 PST 2015
[  116.858245] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20141107/nsarguments-95)
[  116.858321] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20141107/nsarguments-95)
[  116.858373] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20141107/nsarguments-95)
[  116.858425] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20141107/nsarguments-95)
[  116.858475] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20141107/nsarguments-95)
[  116.858663] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20141107/nsarguments-95)
[  116.858830] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20141107/nsarguments-95)
[  116.858883] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20141107/nsarguments-95)
[  116.883277] ACPI Error: Field [TBF3] at 524288 exceeds Buffer [NULL] size 262144 (bits) (20141107/dsopcode-236)
[  116.883280] ACPI Error: Method parse/execution failed [\_SB_.PCI0.PEG_.VID_.GETB] (Node ffff8803318b4be0), AE_AML_BUFFER_LIMIT (20141107/psparse-536)
[  116.883285] ACPI Error: Method parse/execution failed [\_SB_.PCI0.PEG_.VID_._ROM] (Node ffff8803318b4bb8), AE_AML_BUFFER_LIMIT (20141107/psparse-536)
[  116.890749] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20141107/nsarguments-95)
[  117.064022] thinkpad_acpi: asked for hotkey mask 0x0070ffbf, but firmware forced it to 0x0070ffbb
[  124.680820] [drm] Module unloaded
[  124.726666] bbswitch: disabling discrete graphics
[  124.726675] ACPI Warning: \_SB_.PCI0.PEG_.VID_._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20141107/nsarguments-95)
[  124.741243] pci_raw_set_power_state: 51 callbacks suppressed
[  124.741247] pci 0000:04:00.0: Refused to change power state, currently in D0
[  124.757917] thinkpad_acpi: EC reports that Thermal Table has changed

**Comment by ArchangeGabriel** [comment]
@smunaut What can tell if it works here? Does /proc/acpi/bbswitch correctly see the results (i.e. OFF with the hack, ON without)?

**Comment by klebed** [comment]
user@lenovo-T440s:~$ cat /proc/acpi/bbswitch
0000:04:00.0 OFF

*here I've started glxgears in another terminal
user@lenovo-T440s:~$ cat /proc/acpi/bbswitch
0000:04:00.0 ON
user@lenovo-T440s:~$ cat /proc/acpi/bbswitch
0000:04:00.0 ON

*here I've stopped glxgears in another terminal
user@lenovo-T440s:~$ cat /proc/acpi/bbswitch
0000:04:00.0 OFF

And then having some tee:

user@lenovo-T440s:~$ sudo tee /proc/acpi/bbswitch <<<ON
ON
user@lenovo-T440s:~$ cat /proc/acpi/bbswitch
0000:04:00.0 ON
user@lenovo-T440s:~$ sudo tee /proc/acpi/bbswitch <<<OFF
OFF
user@lenovo-T440s:~$ cat /proc/acpi/bbswitch
0000:04:00.0 OFF

**Comment by smunaut** [comment]
Having lspci show "rev ff" for the card has been the best indicator of a true OFF state for me.

**Comment by ArchangeGabriel** [comment]
OK, nice to see. So at least they are some progress for you here.

**Comment by ArchangeGabriel** [comment]
@smunaut That’s what bbswitch is looking at normally.

**Comment by klebed** [comment]
Yep, and idle temperature dropped to +44.0°C approx. Seems like we reduced some carbon dioxide emissions here. :)

**Comment by ivanershov51** [comment]
Sorry, I don't understand a thing, but is my problem related to this bug? The problem: optirun enables nvidia card, but fails to disable it later. Enabling/disabling nvidia manually works, though.
Arch linux, latest kernel. I have acpi_osi= (it's blank after the '=') for my laptop brightness keys to work.

**Comment by emanuil-tolev** [comment]
And you have a Lenovo Thinkpad?

**Comment by ivanershov51** [comment]
@emanuil-tolev: I have ASUS N550JV.

**Comment by emanuil-tolev** [comment]
Hmm. It is possible you would be affected by the same ACPI update problems which plagued Thinkpad T440s and T450s (and presumably their cousin T440, T450 and who knows how many more Lenovo series). If you don't have the time to get into whether your machine had the same problem as these Thinkpads, I would attempt following the instructions on the forked branch: https://github.com/smunaut/bbswitch/tree/hack-t440s and checking dmesg | grep bbswitch.

Actually, you're saying optirun fails to disable your card "later" - do you see evidence of the turning off process failing if you do dmesg | grep bbswitch after using optirun?

Notice this part of the original post:

> Using vanilla bbswith : Switching OFF does put it to D3, but it's not entirely off. The power usage is lower but not as low as it should be (in my case, about 1W lower than full power while it should be more like 2.5W). Also, bbswith reports the card as "ON" because lspci still works fine and you can still access the config space. There is no way to turn the card back ON and so loading nvidia driver will actually crash trying to access a card that's only half powered.

If you are using vanilla bbswitch (e.g. you installed bumblebee from your Arch repos and the package maintainer is not using @smunaut's fork to compile the Arch version, but rather just using Bumblebee-Project/bbswitch like Ubuntu does), then you should see *those* symptoms specifically. It sounds like it's possibly the same problem as "your card stays on" - I presume you mean /cat/proc/bbswitch says "ON" after optirun has finished. Trying it out and seeing it go "OFF" would confirm this.

**Comment by ivanershov51** [comment]
"do you see evidence of the turning off process failing if you do dmesg | grep bbswitch after using optirun?" - Yes. I mean before running "optirun firefox" dmesg | grep bbswitch says "disabling graphics card". When this instance of Frefox is running, dmesg says "enabling graphics card". I close firefox, but the last line of dmesg is "enabling..." and /proc/acpi/bbswitch prints ON, not OFF.

**Comment by morj** [comment]
@smunaut hello! Any chance to have your patched repository back until we have this fix?

**Comment by smunaut** [comment]
!?!? WTF ... I never deleted that repo.

**Comment by smunaut** [comment]
Does anyone have a copy ? I don't have a local checkout at all ...

**Comment by doudou** [comment]
'disable_root_port' branch at https://github.com/doudou/bbswitch

**Comment by smunaut** [comment]
tx !

**Comment by valneacsu** [comment]
Ubuntu 14.04.4 now ships with Kernel 4.2, which means the previous fix (acpi_osi='!Windows 2013') no longer works. Are we any closer to releasing a fix for this issue?

**Comment by valneacsu** [comment]
Replying to myself so others don't have to search. For Kernel 4.2 all I had to do was to replace acpi_osi='!Windows 2013' with acpi_osi=Linux in the GRUB boot parameters.

**Comment by NikolausDemmel** [comment]
@valneacsu: Could you quickly summarize what exactly "works" for you with `acpi_osi=Linux` and what else you needed to do? Do you use the patched bbswitch from https://github.com/doudou/bbswitch/tree/disable_root_port ?

**Comment by valneacsu** [comment]
Sorry, I did not mention that I am using bbswitch v0.8 from the [Bumblebee PPA](https://launchpad.net/~bumblebee/+archive/ubuntu/stable). Other than that I did not have to do anything special, but I've also been using nvidia-355 for some time.

**Comment by smunaut** [comment]
If you make Linux not advertise any "Windows" string to ACPI (using acpi_osi=Linux) then the disable_root_port shouldn't be necessary. It basically uses the "old" switch off method path in the ACPI bios code.

**Comment by NikolausDemmel** [comment]
Ah ok, I didn't get that yet. And with that you can turn the card ON and OFF and it also turns off completely?

**Comment by valneacsu** [comment]
I haven't measured the power consumption, but both /proc/acpi/bbswitch and 'lspci' show the card as off.

**Comment by NikolausDemmel** [comment]
Thanks for the explanations!

**Comment by wuqso** [comment]
I have a Lenovo T450s with Fedora 23.

```
#uname -r
4.4.6-300.fc23.x86_64
```

I downloaded doudou's bbswitch code from https://github.com/doudou/bbswitch, build it with 'make' and load it with 'make load' with no error. But the Nvidia card is still on.

```
# cat /proc/acpi/bbswitch
0000:04:00.0 ON
# tee /proc/acpi/bbswitch <<< OFF
OFF
# cat /proc/acpi/bbswitch
0000:04:00.0 ON
# lspci -vnn | grep '\''[030[02]\]'
00:02.0 VGA compatible controller [0300]: Intel Corporation Broadwell-U Integrated Graphics [8086:1616] (rev 09) (prog-if 00 [VGA controller])
04:00.0 3D controller [0302]: NVIDIA Corporation GM108M [GeForce 940M] [10de:1347] (rev a2)
```
What's the problem? I'm a user rather a programmer. I just want a long battery life in linux system.

**Comment by wuqso** [comment]
@valneacsu, using acpi_osi=Linux solves my problem. Thank you!

**Comment by Lekensteyn** [comment]
For those who are wondering why the methods are different depending on the kernel version (or the setting of `acpi_osi`), apparently Windows 10 (`Windows 2015`) drops the use of DSM completely. For that OS you have to toggle the power resources (basically the parent PCIe port).

David Airlie posted a patch to handle the power resources for nouveau/vgaswitcheroo, but I don't know what happened to those patches (https://lkml.org/lkml/2016/3/9/65). You could probably check the power resources of the parent device, but that smells hacky if you call it directly (shouldn't the Linux PM take care of this?). See also the analysis at https://github.com/Bumblebee-Project/bbswitch/issues/115#issuecomment-218622306. While turning off has no DSM methods, surprisingly there are some involved with turning it on. Can you reproduce/match this with your machines?

(Will comment later on the patch, my battery is dieing)

**Comment by doudou** [comment]
David Airlie's patch definitely looks interesting. Even more so because the PM handler is exported, and could therefore be used outside of vgaswitcheroo-enabled drivers.

**Comment by doudou** [comment]
If you look at the LKML thread, it seems that the kernel devs are tackling the root problem ... that is adding runtime PM to the PCIe root ports. What it means for the future of bbswitch, I'm not sure, since it will work only if the card itself is put in D3 first.

> While turning off has no DSM methods, surprisingly there are some involved with turning it on. Can you reproduce/match this with your machines?

On mine, the NVidia suspend/resume with pcie-root-port in D3 works without any calls to any DSM.

**Comment by Lekensteyn** [comment]
I read the full thread (and some of the linked patchwork entries). David's patch adds a function to enabble PM operations that power off/on the parent (PCIe port) device and hooks it into nouveau. Rafael [commented][2] that the patches PM for PCIe ports are still under discussion. Especially note:

> > I'm guessing on Windows this all happens automatically.
> 
> PCIe ports are power-managend by (newer) Windows AFAICS, but we know for a fact
> that this simply doesn't work reliably on some older hardware which is why
> we don't do that.  I suppose that the Windows in question uses a cut-off date
> or similar to decide what do do with PCIe ports PM.

As for bbswitch's future, it is currently broken for newer devices so I'll probably do this:

 - Add detection for newer devices (TBD, maybe I can check for the `_PR3` method on parent device (PCIe port)? Calling DSM could also be done, but I am afraid that this is less well tested.)
 - If an old device is detected, keep the current situation (important, I don't want to break old devices).
- If a new device is detected:
  - Skip DSM before PS3
  - After PS3, disable the parent device (new; but has to be future-proof as well)

In case upstream Linux adds proper PM to the PCIe root port, then bbswitch should not interfere with that, so there also needs to be checks for that (PM domains?). Some reading material: https://www.kernel.org/doc/Documentation/power/devices.txt

bbswitch differentiates itself from nouveau in that it will always keep the device powered off unless explicitly asked by the user (via /proc/acpi/bbswitch). Currently nouveau will flip the card on when you execute `lspci` for example or open `/dev/dri/...`. Note that I will probably use nouveau though since external screens are connected via the Nvidia card on my laptop.

> On mine, the NVidia suspend/resume with pcie-root-port in D3 works without any calls to any DSM.

Did you observe this on Windows? Is Windows doing any DSM calls after D0?

 [2]: https://lkml.org/lkml/2016/3/10/542

**Comment by doudou** [comment]
> Currently nouveau will flip the card on when you execute lspci for example or open /dev/dri/....

This is due to the usage of runtime PM. lspci triggers a wakeup. I had a version of bbswitch that was purely relying on runtime PM and noticed that.

Once all the PCIe root port work is in the kernel, I'm planning try to use nouveau instead of bbswitch just for the PM work. I still rely on the nvidia proprietary driver for proper opengl support.

> Did you observe this on Windows?

I would not even begin to know how I could check that.

**Comment by shadoxx** [comment]
Just wanted to confirm this issue on a Thinkpad W550s running BIOS version 1.14. `bbswitch` worked fine on the BIOS that was shipped with this laptop which was version 1.06. I would downgrade the BIOS, but Lenovo has disabled that functionality as well, citing security concerns.

**Comment by Lekensteyn** [comment]
@shadoxx https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1452979 lists acpidump for a W550s (Quadro K620M, BIOS 1.02). The SSDT changes are not relevant. DSDT has changes a bit, mainly wrt USB 3.0, some TPM changes. These are the hunks relevant for graphics (GPON, GPOF and Win10 detection):
```diff
diff --git a/acpi-1.02/dsdt.dsl b/acpi-1.14/dsdt.dsl
index 53b2c77..9717867 100644
--- a/acpi-1.02/dsdt.dsl
+++ b/acpi-1.14/dsdt.dsl
@@ -1051,24 +1051,30 @@ DefinitionBlock ("dsdt.aml", "DSDT", 1, "LENOVO", "TP-N11  ", 0x00001020)
                 If (\_OSI ("Windows 2012"))
                 {
                     \WIN8 = 0x01
                     OSYS = 0x07DC
                 }
 
                 If (\_OSI ("Windows 2013"))
                 {
                     \WIN8 = 0x01
                     OSYS = 0x07DD
                 }
 
+                If (\_OSI ("Windows 2015"))
+                {
+                    \WIN8 = 0x01
+                    OSYS = 0x07DF
+                }
+
                 If (\_OSI ("Linux"))
                 {
                     \LNUX = 0x01
                     OSYS = 0x03E8
                 }
 
                 If (\_OSI ("FreeBSD"))
                 {
                     \LNUX = 0x01
                     OSYS = 0x03E8
                 }
             }
@@ -6774,32 +6770,36 @@ DefinitionBlock ("dsdt.aml", "DSDT", 1, "LENOVO", "TP-N11  ", 0x00001020)
                         {
                             GPOF (0x00)
                         }
                     }
 
                     Method (GPON, 1, NotSerialized)
                     {
                         If (ISOP ())
                         {
                             If (DGOS)
                             {
                                 \VHYB (0x02, 0x00)
-                                Sleep (0x64)
+                                Sleep (0x14)
                                 If ((ToInteger (Arg0) == 0x00)) {}
                                 \VHYB (0x00, 0x01)
-                                Local0 = 0x00
-                                While ((Local0 < 0x5A))
+                                Sleep (0x14)
+                                Local2 = \VHYB (0x0E, 0x00)
+                                While ((Local2 != 0x0F))
                                 {
-                                    Local0 += One
-                                    Stall (0x64)
+                                    \VHYB (0x00, 0x00)
+                                    Sleep (0x14)
+                                    \VHYB (0x00, 0x01)
+                                    Sleep (0x0A)
+                                    Local2 = \VHYB (0x0E, 0x00)
                                 }
 
                                 \VHYB (0x02, 0x01)
                                 Sleep (0x01)
                                 \VHYB (0x08, 0x01)
                                 Local0 = 0x0A
                                 Local1 = 0x32
                                 LREN = LTRS /* \_SB_.PCI0.PEG_.LTRS */
                                 CEDR = One
                                 While (Local1)
                                 {
                                     Sleep (Local0)
@@ -6851,31 +6851,25 @@ DefinitionBlock ("dsdt.aml", "DSDT", 1, "LENOVO", "TP-N11  ", 0x00001020)
 
                     Method (GPOF, 1, NotSerialized)
                     {
                         If (ISOP ())
                         {
                             If ((VMSH || (\_SB.PCI0.PEG.VID.OMPR == 0x03)))
                             {
                                 LTRS = LREN /* \_SB_.PCI0.PEG_.LREN */
                                 \SWTT (0x00)
                                 \VHYB (0x08, 0x00)
                                 \VHYB (0x08, 0x02)
                                 \VHYB (0x02, 0x00)
-                                Local0 = 0x00
-                                While ((Local0 < 0x1E))
-                                {
-                                    Local0 += One
-                                    Stall (0x64)
-                                }
-
+                                Sleep (0x09)
                                 \VHYB (0x00, 0x00)
                                 If ((ToInteger (Arg0) == 0x00)) {}
                                 DGOS = One
                                 \_SB.PCI0.PEG.VID.OMPR = 0x02
                             }
                         }
                     }
 
                     Method (_STA, 0, NotSerialized)  // _STA: Status
                     {
                         Return (0x0F)
                     }
```

The ACPI changes do not look very significant, it is mainly the timing that are changed. GPOF went from 30×100µs(=3ms) to 9ms. GPON is slightly more interesting. It changes the timings and adds a SMI call (unknown function). Maybe there were other non-ACPI tunings.

Anyway, maybe things get better when the power resources are used instead of DSM.

**Comment by Lekensteyn** [comment]
Could someone try some patch series on top of v4.7 with nouveau? See https://github.com/Bumblebee-Project/bbswitch/issues/78#issuecomment-223072012

**Comment by Lekensteyn** [comment]
FYI, this has been fixed for the nouveau in Linux v4.8-rc1, bbswitch still needs an update though.

**Comment by BernardoGO** [comment]
@Lekensteyn 
Does it fix the suspension problem for Clevos?
BBswitch currently does not support graphic switch using nouveau, right?

**Comment by Lekensteyn** [comment]
@BernardoGO What suspend problem? System sleep or runtime suspend?
With nouveau, you do not need bbswitch as it is able to handle PM.

**Comment by klebed** [comment]
Just to leave another footprint in this epic issue.

If somebody using ubuntu 16.04 (and probably other distros with >4.4.X), and used nvidia-361, it was force-replaced with nvidia-367 (you install 361, but apt gives you 367 with 361 simultaneously), so you could experience all problems again. First of all, use up to date bumblebee and primus from bumblebee/testing ppa, install nvidia-367. And then, after it started to work, check bbswitch, since it probably stopped working properly. 
The solution is to edit /etc/default/grub , adding acpi_osi=Linux to GRUB_CMDLINE_LINUX parameter, and then do: sudo update-grub .

Worked with Lenovo T440s.

**Comment by RoadToDream** [comment]
Thanks to @smunaut and @klebed . Worked for me.

**Comment by Lacrymology** [comment]
just an update, on a ThinkPad T450s, the card is constantly on and the grub fix above doesn't work (plasma never finishes loading and a bunch of error messages show up in the logs).

**Comment by wuqso** [comment]
Dear Tomas Neme,


Thank you for your email. I have already successfully set up bumblebee on the T450s.


Best,
Jianghua Wu



-----原始邮件-----
发件人:"Tomas Neme" <notifications@github.com>
发送时间:2019-04-25 19:55:10 (星期四)
收件人: Bumblebee-Project/bbswitch <bbswitch@noreply.github.com>
抄送: wuqso <jhwu@bnu.edu.cn>, Comment <comment@noreply.github.com>
主题: [SPAM] Re: [Bumblebee-Project/bbswitch] T440s with kernel >= 3.15 doesn't power off properly (Analysis and possible solution included !) (#112)



just an update, on a ThinkPad T450s, the card is constantly on and the grub fix above doesn't work (plasma never finishes loading and a bunch of error messages show up in the logs).
